adler32.h

```diff
@@ -0,0 +1,84 @@
+#include <stdint.h>
+
+/* adler32 */
+typedef uint32_t (*adler32_func)(uint32_t adler, const uint8_t *buf, size_t len);
+
+extern uint32_t adler32_c(uint32_t adler, const uint8_t *buf, size_t len);
+#define ADLER32 adler32_c
+#ifdef ARM_NEON
+extern uint32_t adler32_neon(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_neon
+#endif
+#ifdef PPC_VMX
+extern uint32_t adler32_vmx(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_vmx
+#endif
+#ifdef RISCV_RVV
+extern uint32_t adler32_rvv(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_rvv
+#endif
+#ifdef X86_SSSE3
+extern uint32_t adler32_ssse3(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_ssse3
+#endif
+#ifdef X86_AVX2
+extern uint32_t adler32_avx2(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_avx2
+#endif
+#ifdef X86_AVX512
+extern uint32_t adler32_avx512(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_avx512
+#endif
+#ifdef X86_AVX512VNNI
+extern uint32_t adler32_avx512_vnni(uint32_t adler, const uint8_t *buf, size_t len);
+#undef ADLER32
+#define ADLER32 adler32_avx512_vnni
+#endif
```

Comment:
If these were from fastest to slowest instead of slowest to fastest, you could eliminate all the `#undef` lines...

Also `adler32_c` should be the last, as fallback if `ADLER32` isn't defined to anything else.

Reply:
I don't think that is the case. Maybe if it were `#ifdef/#elif` but multiple variants have to be available if the compiler supports it.

Reply:
Hmm. Let me play around with it and see if I can clean these up a bit.

Reply:
I've tried to clean it up, but in the end it is always a trade off between doing `#undef X` or `#ifndef X/#endif`. And `#undef X` just is just more straightforward imo.